File: mobile/test/drift_test.dart
// Drift check: loads the repo-wide golden fixture and asserts the Dart wire
// codec roundtrips each case. Paths are resolved relative to the mobile/
// package root because `flutter test` sets that as CWD.
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:appunvs_mobile/pb/wire.dart';

Object? _sort(Object? v) {
  if (v is Map<String, dynamic>) {
    final List<String> keys = v.keys.toList()..sort();
    return <String, dynamic>{
      for (final String k in keys) k: _sort(v[k]),
    };
  }
  if (v is Map) {
    // Map<String, Object?> or similar
    final Map<String, Object?> m =
        v.map((Object? k, Object? val) => MapEntry<String, Object?>(k.toString(), val));
    return _sort(m);
  }
  if (v is List) return v.map(_sort).toList();
  return v;
}

void main() {
  final File golden = File('../shared/proto/testdata/messages.json');
  if (!golden.existsSync()) {
    throw StateError('golden fixture missing at ${golden.absolute.path}');
  }
  final List<dynamic> cases =
      jsonDecode(golden.readAsStringSync()) as List<dynamic>;
  if (cases.isEmpty) {
    throw StateError('golden fixture is empty');
  }

  group('wire drift against shared/proto/testdata', () {
    for (final dynamic c in cases) {
      final Map<String, dynamic> entry = c as Map<String, dynamic>;
      final String name = entry['name'] as String;
      final Map<String, dynamic> msg = entry['message'] as Map<String, dynamic>;

      test(name, () {
        final Message parsed = Message.fromJson(msg);
        final Map<String, dynamic> produced = parsed.toJson();
        expect(_sort(produced), equals(_sort(msg)));
      });
    }
  });
}

File: browser/src/lib/pb/drift.test.ts
// Drift check: loads the repo-wide golden fixture and asserts the TS wire
// codec roundtrips each case to byte-identical canonical JSON. If this test
// fails the TS implementation has drifted from shared/proto/testdata.
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { fromJson, toJson } from './wire';

interface Case {
  name: string;
  message: Record<string, unknown>;
}

function loadCases(): Case[] {
  const p = resolve(__dirname, '../../../../shared/proto/testdata/messages.json');
  return JSON.parse(readFileSync(p, 'utf8')) as Case[];
}

function canonical(obj: unknown): string {
  // Deep sort keys so field order never matters.
  const sort = (v: unknown): unknown => {
    if (v === null || typeof v !== 'object') return v;
    if (Array.isArray(v)) return v.map(sort);
    return Object.keys(v as Record<string, unknown>)
      .sort()
      .reduce<Record<string, unknown>>((acc, k) => {
        acc[k] = sort((v as Record<string, unknown>)[k]);
        return acc;
      }, {});
  };
  return JSON.stringify(sort(obj));
}

describe('wire drift against shared/proto/testdata', () => {
  const cases = loadCases();
  expect(cases.length).toBeGreaterThan(0);

  for (const c of cases) {
    it(c.name, () => {
      const goldenJson = JSON.stringify(c.message);
      const parsed = fromJson(goldenJson);
      const produced = toJson(parsed);
      expect(canonical(JSON.parse(produced))).toBe(canonical(c.message));
    });
  }
});
